COR-569: Support For Non-Image View Rendering & Paperclip Non-Image Upload bug fix

MKwenhua opened the pull request:
**EDIT:**
I have decided that instead of removing the styles attribute, I just nested the data under _image_styles_:
```ruby 
metadata:
 {
  image_styles: {
      styles: {
        large: {geometry: '1800x1800>', format: :jpg},
        medium: {geometry: '800x800>', format: :jpg},
        default: {geometry: '300x300>', format: :jpg},
        mini: {geometry: '100x100>', format: :jpg},
        micro: {geometry: '50x50>', format: :jpg},
        post_tile: {geometry: '1140x', format: :jpg}
      },
      processors: [:thumbnail, :paperclip_optimizer]
   },
   preserve_files: true,
   path: ':class/:attachment/careerbuilder-:style-:id.:extension',
   s3_headers: {'Cache-Control': 'public, max-age=315576000'}
}
```
This allows users to still define styles without having Paperclip apply these styles to non-image assets. 

Additionally in [cortex pr#421](https://github.com/cbdr/cortex/pull/421) I pass in `ActionDispatch` data with all metadata under `:field_data`:
```ruby
  def format_data(metadata_hash)
   return metadata_hash unless is_image?(metadata_hash[:field_data]["asset"].content_type)
   metadata_hash.merge(metadata_hash["image_styles"])
  end
```


**--Outdated changes below---**

I removed `styles` and `processors` from `media.rake` and moved that code into a method as seen below: 
```ruby 
def image_asset_styles
  {
    styles: {
      large: {geometry: '1800x1800>', format: :jpg},
      medium: {geometry: '800x800>', format: :jpg},
      default: {geometry: '300x300>', format: :jpg},
      mini: {geometry: '100x100>', format: :jpg},
      micro: {geometry: '50x50>', format: :jpg},
      post_tile: {geometry: '1140x', format: :jpg}
    },
    processors: [:thumbnail, :paperclip_optimizer]
  }
end
```

which is used to merge into `@metadata` if the assets content_type is an image/pdf: 

```ruby
def format_data(metadata_hash)
  return metadata_hash unless is_image(metadata_hash[:content_type])
  metadata_hash.merge(image_asset_styles)
end
```

this relies on content_type being passed to `AssetFieldType` in cortex(which is a separate PR). 

Additionally I added `application/zip` to **allowed_asset_content_types**. Another small bug I had to fix was `asset_info` index and show were throwing errors expecting an asset thumbnail and reading dimensions .

ElliottAYoung commented on app/models/asset_field_type.rb:
```diff
@@ -52,6 +53,29 @@ def mapping
+
+  def image_asset_styles
+    {
+      styles: {
```
So part of the reason that styles existed in media.rake was that it allowed the User to specify which size uploads they want, which does vary from case to case (especially things like the names of the styles - post_tile is a pretty CB specific concern).

MKwenhua replied:
Ok, then maybe we should have an array of image sizes and their dimensions that isn't set with the key `styles`, we could even change it to `image_asset_styles` which we could pass as styles to Paperclip in case it is an image. When Paperclip sees that there is a styles object, it will pass whatever the file is to ImageMagic no matter what.

ElliottAYoung commented on app/models/asset_field_type.rb:
```diff
@@ -52,6 +53,29 @@ def mapping
 
   private
 
+  def is_image(mime_type)
```
How does this method differ from the `image?` method below on lines 55-56?

ElliottAYoung replied:
Also - when writing a boolean method, please add `?` after the name for self-documentation on what it returns

MKwenhua replied:
The difference is in `image?` compares against `asset_content_type` which is not yet defined when I need it

```ruby
def image?
  asset_content_type =~ %r{^(image|(x-)?application)/(bmp|gif|jpeg|jpg|pjpeg|png|x-png)$}
end
```

also `image?` returns either `nil` or `0`  and I hoping for a true boolean method

ElliottAYoung commented on app/models/asset_field_type.rb:
```diff
@@ -52,6 +53,29 @@ def mapping
+        micro: {geometry: '50x50>', format: :jpg},
+        post_tile: {geometry: '1140x', format: :jpg}
+      },
+      processors: [:thumbnail, :paperclip_optimizer]
```
Again, not every user that would touch this would necessarily want thumbnails with all image assets. As far as assumptions go it's not the worst we could make, but we still want to avoid assuming anything about future User's needs / want to leave everything as flexible and customizable as possible

MKwenhua replied:
ok then how about I just move styles into a key like `image_styles` in **media.rake**

```ruby
image_styles: {
  styles: {
    large: {geometry: '1800x1800>', format: :jpg},
    medium: {geometry: '800x800>', format: :jpg},
    default: {geometry: '300x300>', format: :jpg},
    mini: {geometry: '100x100>', format: :jpg},
    micro: {geometry: '50x50>', format: :jpg},
    post_tile: {geometry: '1140x', format: :jpg}
  },
  processors: [:thumbnail, :paperclip_optimizer]
},
preserve_files: true,
path: ':class/:attachment/careerbuilder-:style-:id.:extension',
s3_headers: {'Cache-Control': 'public, max-age=315576000'}
```

ElliottAYoung commented on app/models/asset_field_type.rb:
```diff
@@ -34,6 +34,7 @@ def data
         'dimensions': dimensions,
         'content_type': asset_content_type,
         'file_size': asset_file_size,
+        'has_thumb': is_image(asset_content_type),
```
Again, this is assuming that all image media file should have thumbnails attached, which may not necessarily be the case

MKwenhua replied:
Ok but in asset_info that was the preexisting assumption 

```ruby
- if asset
  = image_tag(asset['style_urls'][config[:thumbnail_style]], height: '50px')
```

additionally this is also used in Show where an `image_tag` is rendered as well, I made this function to and image element wont have a docx file as a source

ElliottAYoung commented on lib/tasks/cortex/core/media.rake:
```diff
@@ -29,15 +29,7 @@ namespace :cortex do
-                               post_tile: {geometry: '1140x', format: :jpg}
-                             },
-                             processors: [:thumbnail, :paperclip_optimizer],
+                             styles: {},
```
If styles have been cut out of rake task and hardcoded why have this at all?

MKwenhua replied:
Other methods assume styles does exist, which would throw an error otherwise, which is why we should probably have a helper function.I made one for that React cells demo, as well as a in browser text editor as well.

This would allow any config that the user creates to not break.

toastercup commented on app/models/asset_field_type.rb:
```diff
@@ -52,6 +53,29 @@ def mapping
+  end
+
+  def format_data(metadata_hash)
+   return metadata_hash unless is_image(metadata_hash[:content_type])
```
As a potential solution to Elliott's concerns, couldn't you just remove the `:thumbnail` and `:paperclip_optimizer` strings from the `processors` array in this method?

MKwenhua replied:
I did that in a previous version, but I felt it was messier than adding processors later.

arelia reviewed (commented):
:cool:

MKwenhua commented:
@toastercup  @ElliottAYoung I just wanted to point out that even though we will instead be switching to  [github.com/janko-m/shrine](https://github.com/janko-m/shrine)  [as noted in cortex/pull #421](https://github.com/cbdr/cortex/pull/421) we should still account for rendering bugs for non-image asset content items show and index thumbnails. I can create a separate PR with just those changes if you'd like.

toastercup commented on lib/tasks/cortex/core/media.rake:
```diff
@@ -17,7 +17,7 @@ namespace :cortex do
         puts "Creating Fields..."
 
-        allowed_asset_content_types = %w(txt css js pdf doc docx ppt pptx csv xls xlsx svg ico png jpg gif bmp)
+        allowed_asset_content_types = %w(txt css js pdf doc docx ppt pptx csv xls xlsx zip odt epub svg ico png jpg gif bmp)
```
ZIP should not be supported

MKwenhua replied:
Oh yeah, I think the reason why I added zip has to do with how Paperclip reads `docx` files created on google docs since i think it has some zipped/compressed config docs contained within it. 

But actually I don't think it's necessary for this to work, so I can remove it.

MKwenhua committed:
changed styles to image_styles in media.rake to allow user defined image styles

changed styles to image_styles in media.rake to allow user defined image styles

yyy

changed styles to image_styles in media.rake to allow user defined image styles

changed styles to image_styles in media.rake to allow user defined image styles

changed styles to image_styles in media.rake to allow user defined image styles

replaced is_image? method with MimeType

replaced is_image? method with MimeType and added image boolean to data

replaced is_image? method with MimeType and added image boolean to data

toastercup committed:
Add support for width, height, style and class; always rebuild these attributes on data init for CKEditor Media Widget

toastercup commented on app/models/asset_field_type.rb:
```diff
@@ -52,6 +52,15 @@ def mapping
   private
 
+  def is_image?(mime_type)
+    ['image/jpeg', 'image/pjpeg', 'image/png','application/pdf' ,'image/x-png', 'image/gif'].include?(mime_type)
```
`MimeMagic` gives you a method that allows you to infer what sort of type a mimetype is. Try: `MimeMagic.new(mime_type).mediatype == 'image'`. I'd also add a manual check for PDF. Either way, this `is_image?` method misses on some of our supported filetypes, like BMP.

Additionally, please follow the Ruby convention for predicate methods - this should be `image?` instead of `is_image?`

Edit: I see there are two methods that do similar things. This is definitely where a good name comes in handy - this should be named something more like `can_post_process?` or `has_thumb?`

MKwenhua replied:
I actually made this Method to check in
```ruby
 def metadata=(metadata_hash)
    @metadata = format_data(metadata_hash).deep_symbolize_keys
    @existing_data = metadata_hash[:existing_data]
    Paperclip::HasAttachedFile.define_on(self.class, :asset, existing_metadata)
end
```
To change `image_styles` to `styles` so that if the asset is an image or pdf that it can run post processing. Originally I tried `image?` but that relies on `asset_content_type` being undefined at that moment.

toastercup replied:
Few things: are you replying to the right comment? Is that the correct method you pasted? I'm not sure I understand what you're saying

toastercup commented on app/models/asset_field_type.rb:
```diff
@@ -52,6 +52,15 @@ def mapping
+    ['image/jpeg', 'image/pjpeg', 'image/png','application/pdf' ,'image/x-png', 'image/gif'].include?(mime_type)
+  end
+
+  def format_data(metadata_hash)
```
Please come up with a better name for this method - it isn't just formatting data

toastercup commented on lib/tasks/cortex/core/media.rake:
```diff
@@ -29,15 +29,17 @@ namespace :cortex do
-                               mini: {geometry: '100x100>', format: :jpg},
-                               micro: {geometry: '50x50>', format: :jpg},
-                               post_tile: {geometry: '1140x', format: :jpg}
+                            image_styles: {
```
This breaks from my original intention in designing the syntax for asset config - I wanted it to mimic Paperclip's config almost 1:1. Try keeping the original `styles` key. In the `format_data` method, just delete the `styles` and `processors` hash keys if you need to remove styles for a non-post-processed asset.

MKwenhua replied:
Ok I originally had that in a past commit, but I was told to change it. So what I do instead is I just change `image_styles` to `styles` if it's an image asset

toastercup replied:
I think you were asked to change it because using Paperclip's hook to prevent post processing would have been ideal, but we discovered that wasn't possible. I'm okay with deleting from the original `styles` key and producing a new hash for use in non-image assets.

MKwenhua commented on app/cells/plugins/core/asset_info/show.haml:
```diff
@@ -21,7 +24,7 @@
         %dt Dimensions
         %dd
-          = dimensions
+          = asset['image'] ? dimensions : 'Not Image Asset'
```
I added `image` to the data method to avoid rendering nonexistent data

MKwenhua commented on app/models/asset_field_type.rb:
```diff
@@ -52,6 +53,11 @@ def mapping
 
   private
 
+  def asset_post_processing_data(metadata_hash)
```
I changed the name so that it's no longer `format_data`, let me know if you still want it changed

toastercup commented on app/models/asset_field_type.rb:
```diff
@@ -30,6 +30,7 @@ def data
       'asset': {
         'file_name': asset_file_name,
         'url': asset.url,
+        'image': !!image?,
```
To conform to Ruby conventions, predicate methods must always return true or false - we shouldn't have to cast to a Boolean here

MKwenhua replied:
yes, but this method returns `nil` or `0`, which is why I made the other predicate method `is_image?`.

toastercup commented on app/models/asset_field_type.rb:
```diff
@@ -52,6 +53,11 @@ def mapping
   private
 
+  def asset_post_processing_data(metadata_hash)
+   return metadata_hash unless MimeMagic.new(field_data["asset"].content_type).mediatype == 'image'
```
In more complex methods that you'd want to break out of early, I'm okay with multiple return paths, but this method could easily be re-written with a standard if/else

toastercup commented on app/models/asset_field_type.rb:
```diff
@@ -119,6 +125,7 @@ def validate_asset_content_type
   end
 
   def style_urls
+    return unless image?
```
I'm okay with multiple return paths here - breaking out early makes sense.

MKwenhua replied:
Yeah and it matches `def extract_dimensions` which also uses the `image?` method so I was trying to match style.

toastercup commented on lib/tasks/cortex/core/media.rake:
```diff
@@ -29,15 +29,17 @@ namespace :cortex do
-                               mini: {geometry: '100x100>', format: :jpg},
-                               micro: {geometry: '50x50>', format: :jpg},
-                               post_tile: {geometry: '1140x', format: :jpg}
+                            image_styles: {
```
This will still need fixing up prior to merge

toastercup commented on app/cells/plugins/core/asset_info/show.haml:
```diff
@@ -7,7 +7,10 @@
+        - if asset['image']
+          = image_tag(asset['url'], style: 'max-height: 200px;')
+        - else
+          = asset['content_type']
```
I just wouldn't display anything if there's no thumbnail to display

toastercup commented on app/cells/plugins/core/asset_info/show.haml:
```diff
@@ -21,7 +24,7 @@
         %dt Dimensions
         %dd
-          = dimensions
+          = asset['image'] ? dimensions : 'Not Image Asset'
```
Wrap the whole `dt`/`dd` Dimensions bit in a conditional and don't render if it's not an image

MKwenhua replied:
Ok will do

toastercup commented on app/cells/plugins/core/asset_info/index.haml:
```diff
@@ -1,2 +1,4 @@
-- if asset
+- if asset['image']
```
Please wrap this entire file in `-if asset` again

toastercup commented on app/assets/javascripts/ckeditor/plugins/cortex_media_insert/plugin.js:
```diff
@@ -7,14 +7,22 @@
-          if (this.data.id) {
-            var image_element = this.element.getFirst(),
-              alt_text = this.data.alt || this.data.title;
+          var image_element = this.element.getFirst(),
```
Please rebase with develop so this diff doesn't appear

MKwenhua replied:
Ok

arelia commented:
@MKwenhua @toastercup @ElliottAYoung Can we add a label for "Not Active" for PRs that don't need to be reviewed for a while? Looks like this would be one of them. I think Pete would have to add the label for this one b/c I can't do it

MKwenhua commented:
@arelia I agree but there seems to be no labels for `not active` I will try and add it

arelia commented:
I added the `Not Active` label and when this becomes active again we can remove it!

ElliottAYoung commented:
@MKwenhua What is happening with this PR? Last activity on it was over two months ago

toastercup commented:
@ElliottAYoung I've gone ahead and closed this - I was keeping it open, but now that we're officially switching to Shrine, we won't be needing it. Thanks! cc @MKwenhua
